vowpalwabbit/parser.cc

```diff
@@ -925,35 +963,75 @@ void finish_example(vw& all, example& ec)
     all.example_parser->output_done.notify_one();
   }
 }
+
+void finish_example_vector(vw& all, std::vector<example*>& ev)
+{
+  ev.clear();
+  all.example_parser->example_vector_pool.return_object(&ev);
+}
 }  // namespace VW
 
-void thread_dispatch(vw& all, const v_array<example*>& examples)
+void thread_dispatch(vw& all, const std::vector<example*>& examples)
 {
-  all.example_parser->end_parsed_examples += examples.size();
-  notify_examples_cv(examples[0]);
+  for (auto example : examples) notify_examples_cv(example);
 }
 
 void main_parse_loop(vw* all) { parse_dispatch(*all, thread_dispatch); }
 
 namespace VW
 {
-example* get_example(parser* p) { 
+std::vector<example*>* get_example(parser* p) { 
 
-  example* ex = p->ready_parsed_examples.pop();
+  std::vector<example*>* ev = p->ready_parsed_examples.pop();
 
-  if (ex == nullptr) {
-    return ex;
+  if (ev == nullptr) {
+    return ev;
   }
 
+  // Frankly, we can take any example in the vector of examples. If one of them is parsed, we can rest assured that all are.
+  example* ex = (*ev)[0];
   {
     std::unique_lock<std::mutex> lock(*(ex->ex_lock.example_cv_mutex));
-    while(ex != nullptr && !*(ex->ex_lock.done_parsing)) {
+    while(ev != nullptr && !*(ex->ex_lock.done_parsing)) {
       ex->ex_lock.example_parsed->wait(lock);
     }
   }
 
-  return ex;
+  return ev;
+
+}
+
+void work_on_example(vw& all, example* ex) {
+  // After parsing the example, we set it up, if the example is not end pass.
+  if (!ex->end_pass){
+    VW::setup_example(all, ex);
+  }
+
+  // if it is end_pass, we execute the end pass sequence
+  else{
+    reset_source(all, all.num_bits);
+    // all.do_reset_source = false;
+    all.passes_complete++;
 
+    // setup an end_pass example
+    all.example_parser->lbl_parser.default_label(&ex->l);
+    all.example_parser->in_pass_counter = 0;
+
+    // if (all.passes_complete == all.numpasses && example_number == all.pass_length)
```

Comment:
why are these lines commented out?

Reply:
Taken care of by adding a global example_number :+1: